### PR TITLE
MERGE PROFILE PAGE:

### LINKED ISSUE #56: Error 500 on double usernames
WIP : thinking about duplicate usernames.

Maybe we could add a unique id after each login.

### LINKED ISSUE #66: Add images in database
The subject states that we need to be able to change the avatar to something else.
image

It means that somehow the user should be able to upload a custom image.

The user should also have a default image -> we take the intra profile picture so that's fine.

### PR DESCRIPTION
**Pull request by riblanc**
Add a basic profile page including:
Avatar upload.
Ability to change username.
Add a suffix when username already taken.
Add basic status management.

Close #66 
Close #56

### CODE REVIEW

File: back/.gitignore [hunk 1]
```diff
@@ -33,4 +33,6 @@ lerna-debug.log*
 !.vscode/settings.json
 !.vscode/tasks.json
 !.vscode/launch.json
-!.vscode/extensions.json
\ No newline at end of file
+!.vscode/extensions.json
```
Comment (AdrianWeaver) [inline back/.gitignore:36]: It's only now that I see those files.
I think we don't want these exceptions to be in here :upside_down_face:

File: back/src/auth/auth.module.ts [hunk 2]
```diff
@@ -11,7 +11,7 @@ import { DeviceService } from './device/device.service';
 import { PrismaModule } from '@prisma/prisma.module';
 
 @Module({
-	imports: [UserModule, JwtModule, PrismaModule],
+	imports: [forwardRef(() => UserModule), JwtModule, PrismaModule],
```
Comment (AdrianWeaver) [inline back/src/auth/auth.module.ts:14]: If you have time, could you explicit this syntax please ?

File: back/src/chat/chat.gateway.ts [hunk 3]
```diff
@@ -28,5 +36,11 @@ export class ChatGateway implements OnGatewayInit, OnGatewayConnection {
 		const user = await this.authService.getWsUser(socket);
 		if (!user) socket.disconnect(true);
 		else await this.chatService.onConnection(socket);
+		this.chatService.emit(`user:${user.id}:status`, await this.statusService.getByUserId(user.id));
+	}
+
+	async handleDisconnect(socket: Socket): Promise<void> {
+		if (!socket.user) return;
+		this.chatService.emit(`user:${socket.user.id}:status`, await this.statusService.getByUserId(socket.user.id));
```
Comment (AdrianWeaver) [inline back/src/chat/chat.gateway.ts:44]: So you also made that to prepare the field for online status, right?

File: back/src/filters/serve-static.exception.ts [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+export class ServeStaticExceptionFilter implements ExceptionFilter {
+	catch(exception: any, host: ArgumentsHost): any {
+		const ctx = host.switchToHttp();
+		const response = ctx.getResponse<Response>();
+		const status = exception instanceof HttpException ? exception.getStatus() : HttpStatus.INTERNAL_SERVER_ERROR;
+
+		if (exception.code === 'ENOENT') response.status(HttpStatus.NOT_FOUND).json(exception);
+		response.status(status).json(exception.response);
+	}
+}
```
Comment (AdrianWeaver) [inline back/src/filters/serve-static.exception.ts:14]: I'm really sorry but I have no idea what this is made for, if you could also explicit please.
(we can do that when you're here) so you don't need to type everything.

File: back/src/types/user.dto.ts [hunk 5]
```diff
@@ -1,5 +1,12 @@
-import { IsNumber, IsString, IsNotEmpty } from 'class-validator';
+import { IsNumber, IsString, IsNotEmpty, Length, Matches } from 'class-validator';
 
+export class UsernameDto {
+	@IsNotEmpty({ message: 'Username is required' })
+	@IsString({ message: 'Username must be a string' })
+	@Length(3, 32, { message: 'Username must be between 3 and 32 characters' })
+	@Matches(/^[a-zA-Z0-9_]+$/, { message: 'Username must only contain alphanumeric characters and underscores' })
+	username: string;
```
Comment (AdrianWeaver) [inline back/src/types/user.dto.ts:8]: You need to allow ``-`` as well.
Some 42 usernames contain hyphens.

File: back/src/user/user.service.ts [hunk 6]
```diff
@@ -29,12 +29,17 @@ export class UserService {
...
+		return this.prisma.user
+			.upsert({
+				where: { id: data.id },
+				create: data,
+				update: {}
+			})
+			.catch(async (e) => {
+				if (e.code === 'P2002') {
+					return this.createOrUpdate({ ...data, username: data.username + '_' + Math.floor(Math.random() * 10000) });
+				}
```
Comment (AdrianWeaver) [inline back/src/user/user.service.ts:41]: I'm not sure that I understand this part.

**Review by AdrianWeaver — COMMENTED** [review]
Hey I asked a few questions on this PR. It's mostly in order to understand what is done even though it seems to be working as expected.
If you have time to explain here or when we're live I would appreciate.
I did not review the front files as I do not feel at ease to do it.

File: back/src/chat/chat.gateway.ts (hunk 3 above)
Comment (riblanc) [reply back/src/chat/chat.gateway.ts:47]: Yeah, in fact i implemented a basic status management in this PR.
For now, the offline and online status are already implemented and working.

File: back/src/auth/auth.module.ts (hunk 2 above)
Comment (riblanc) [reply back/src/auth/auth.module.ts:14]: It's only for circular dependencies purpose
As AuthModule import UserModule which import StatusModule which import ChatModule which import AuthModule etc.
The forwardRef just permit to break the circular depency between several modules

File: back/.gitignore (hunk 1 above)
Comment (riblanc) [reply back/.gitignore]: woops :eyes:

File: back/src/filters/serve-static.exception.ts (hunk 4 above)
Comment (riblanc) [reply back/src/filters/serve-static.exception.ts:14]: It permits to catch 500 errors when accessing to non-existent avatar images and return a 404 error instead.

https://github.com/nestjs/serve-static/issues/139

File: back/src/types/user.dto.ts (hunk 5 above)
Comment (riblanc) [reply back/src/types/user.dto.ts]: Done!

File: front/components/profile/avatar.vue [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+  }),
+  methods: {
+    async uploadFile(e: any) {
+      try {
+        console.log(e.target.files[0]);
+        const formData = new FormData();
+        formData.append("file", e.target.files[0]);
+        for (const entry of formData.entries()) {
+          console.log(entry);
+        }
```
Comment (Elena-Lou) [inline front/components/profile/avatar.vue:31]:
```suggestion

```
console.log()

File: front/components/profile/profile.vue [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+    editable() {
+      return this.user && this.user.id === this.$auth.user.id;
+    },
+    uploadFile: async (e: any) => {
+      console.log(typeof e);
+    },
+    onUpdateUser(user: User) {
+      console.log(user);
+      this.user = user;
+    },
```
Comment (Elena-Lou) [inline front/components/profile/profile.vue:42]:
```suggestion

```

Following direct review, these functions are outdated

File: front/components/profile/username.vue [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+            break;
+          case 409:
+            this.set_error('Username already taken');
+            break;
+          default:
+            this.$emit('alert:show', { title: e.data.message, message: e.data.error });
+        }
+      })
+      if (user) {
+        console.log(user);
```
Comment (Elena-Lou) [inline front/components/profile/username.vue:55]:
```suggestion

```

**Review by Elena-Lou — CHANGES_REQUESTED** [review]
Thank you for your quick work on this new feature

**Review by AdrianWeaver — DISMISSED** [review]
I'm approving the back files.
When you're good with the front you can merge @Elena-Lou

File: front/components/profile/profile.vue [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+  async unmounted() {
+    this.socket.off(`user:${this.id}:status`);
+  },
+  methods: {
+    editable() {
+      return this.user && this.user.id === this.$auth.user.id;
+    },
+    uploadFile: async (e: any) => {
+      console.log(typeof e);
+    },
```
Comment (Elena-Lou) [inline front/components/profile/profile.vue:38]:
```suggestion
```

File: front/components/profile/profile.vue [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+    onUpdateUser(user: User) {
+      console.log(user);
```
Comment (Elena-Lou) [inline front/components/profile/profile.vue:40]:
```suggestion

```

**Review by Elena-Lou — CHANGES_REQUESTED** [review]

**Review by Elena-Lou — APPROVED** [review]

**Pull request merged**